modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+          r = f.reader(locking=false, region=taskBounds, deserializer=des),
+          s = new t();
+
+      // read all the 't' values in the chunk into a list
+      while r.read(s) do
+        results[tid].pushBack(s);
+    }
+
+    // gather results from each task into an array
+    const nPerTask = [r in results] r.size,
+          itemOffsets = (+ scan nPerTask) - nPerTask;
+
+    var result: [0..<itemOffsets.last] t;
+    coforall tid in 0..<nTasks do
+      result[itemOffsets[tid]..<itemOffsets[tid+1]] = results[tid].toArray();
+
+    return result;
+  }
+
+  /*
+    Get an array of ``n+1`` byte offsets that divide the file ``f`` into ``n``
+    roughly equally sized chunks, where each byte offset lines up with a
+    delimiter.
+
+    :arg f: the file to search
+    :arg delim: the delimiter to use to separate the file into chunks
+    :arg n: the number of chunks to find
+    :arg bounds: a range of byte offsets to break into chunks
+    :arg header: a header policy to use when searching for the first byte offset
+
+    :returns: a length ``n+1`` array of byte offsets (the last offset is
+              ``bounds.high``)
+
+    :throws: :class:`OffsetNotFoundError` if a valid byte offset cannot be found
+              in any of the chunks
+  */
+  proc findDelimChunks(const ref f: file, in delim: ?dt, n: int, bounds: range,
+                       header = headerPolicy.noHeader
+  ): [] int throws
+    where dt == bytes || dt == string
+  {
+    const (findStart, startOffset) = header.apply(f, bounds),
+          chunkIndices = if findStart then 0..<n else 1..<n;
+
+    const nDataBytes = bounds.high - startOffset,
+          approxBytesPerChunk = nDataBytes / n;
+
+    var chunkOffsets: [0..n] int;
+    chunkOffsets[0] = startOffset;
+    chunkOffsets[n] = bounds.high;
+
+    coforall i in chunkIndices with (ref chunkOffsets) {
+      const estOffset = chunkOffsets[0] + i * approxBytesPerChunk,
+            r = f.reader(locking=false, region=estOffset..);
+
+      try {
+        r.advanceThrough(delim); // advance past the next delimiter
+      } catch {
+        // there wasn't an offset in this chunk
+        throw new OffsetNotFoundError(
+          "Failed to find byte offset in the range (" + bounds.low:string + ".." + bounds.high:string +
+          ") for chunk " + i:string + " of " + n:string + ". Try using fewer tasks."
+        );
+      }
+      chunkOffsets[i] = r.offset(); // record the offset for this chunk
+    }
+
+    return chunkOffsets;
+  }
+
+  /*
+    Get an array of ``n+1`` byte offsets that divide the file ``f`` into ``n``
+    roughly equally sized chunks, where each byte offset lines up with a
+    delimiter and a ``t`` can be deserialized at that offset.
+
+    This procedure is similar to :proc:`findDelimChunks`, except that when it
+    finds a delimiter, it confirms that a ``t`` can be deserialized at that
+    offset before recording it. This way, the serialized values can also
+    contain the delimiter.
```

Comment:
This feels like a heuristic to me. I'm not sure it's fully reliable. I think it might be OK for CSV if we sometimes have things in `""` maybe not for other things.

Suppose I have a format that uses `;` as a separator but `\;` means not a separator.

```
hello \; there
another line ;
```

If we choose to split at the `;` in the `\;` in the first line, presumably deserializing would simply read until `;` so it would find `hello \` and then ` there\nanother line `.

I think this specific case could be resolved by adding some kind of way for a user to write something that checks if the byte previous to the `;` was a `\`, and if so, not consider it a delimiter. This could be achieved by going to a general regex or just with a point solution for delimiters.

Even for `""` it might not always work as you expect

```
1, "hello
", 2
3, "there"
```

If we choose to split at newline, then we might read `1, "` as the first thing. Which won't deserialize right if you specify it must end there, but you don't here. If we are reading an integer and a quoted string, it would read `(1, "hello\n")` just fine, I would presume. Then reading the next record would fail because there is no 1st element that is an integer. But what if the CSV file mixes quoted and unquoted strings and we are reading it all as strings? Then I think this case could get all confused.

IMO the best we can do here is to provide the user with a way to poke around near the candidate separator to decide if it's acceptable. That could include checking to see if the byte before is `\` (for the quoting case) or it could include checking to see if the previous line had an unterminated string literal, for example. I think what is completely correct here isn't even one thing for all CSV files but is unfortunately more context dependent.

Of course, you could include your current technique as one implementation for "poke around near the candidate separator".

Reply:
I agree that there could be other ways of determining whether an offset is a valid starting point for a value, but I think those could be added later on. For now, I think that what I have is okay because the docs clearly state that:
```
``t`` must:
          * have a 'deserialize' method that throws when a ``t`` cannot be read
  ```

Reply:
IMO you should make the heuristic replaceable by users, but that can be put off (since this is not for stabilization). But, if you keep things as they are here, it's critical to communicate in the docs comment that this code is using a heuristic that is not completely reliable.

Reply:
Okay, I think that's a good direction longer term. Maybe that feature could be implemented by accepting an FCF that determines whether a given offset is valid.